osgp/platform/osgp-adapter-domain-core/src/main/java/org/opensmartgridplatform/adapter/domain/core/application/services/FirmwareManagementService.java

```diff
@@ -367,38 +364,148 @@ private void checkFirmwareHistory(final String deviceId,
+        final Map<String, FirmwareVersionWithInstallationDate> currentlyInstalledFirmwareVersionsPerType = new HashMap<>();
+
+        for (final DeviceFirmwareFile firmwareFile : deviceFirmwareFiles) {
+
+            final Map<FirmwareModule, String> fwms = firmwareFile.getFirmwareFile().getModuleVersions();
+            final Date installationDate = firmwareFile.getInstallationDate();
+
+            for (final Map.Entry<FirmwareModule, String> entry : fwms.entrySet()) {
+                final String version = entry.getValue();
+                final FirmwareModule fwm = entry.getKey();
+                // check if this installation of this same kind of module is
+                // of a later date
+                if (currentlyInstalledFirmwareVersionsPerType.containsKey(fwm.getDescription())
+                        && currentlyInstalledFirmwareVersionsPerType.get(fwm.getDescription())
+                                .getInstallationDate()
+                                .before(installationDate)) {
+                    currentlyInstalledFirmwareVersionsPerType.replace(fwm.getDescription(),
+                            new FirmwareVersionWithInstallationDate(installationDate, new FirmwareVersion(
+                                    FirmwareModuleType.forDescription(fwm.getDescription()), version)));
+                } else {
+                    // no other module of this type found yet so just add it
+                    currentlyInstalledFirmwareVersionsPerType.put(fwm.getDescription(),
+                            new FirmwareVersionWithInstallationDate(installationDate, new FirmwareVersion(
+                                    FirmwareModuleType.forDescription(fwm.getDescription()), version)));
+                }
+            }
+        }
+
+        final List<FirmwareVersion> latestfirmwareVersionsOfEachModuleTypeInHistory = currentlyInstalledFirmwareVersionsPerType
+                .values()
+                .stream()
+                .map(e -> new FirmwareVersion(e.getFirmwareVersion().getFirmwareModuleType(),
+                        e.getFirmwareVersion().getVersion()))
+                .collect(Collectors.toList());
+
+        // remove the latest history (module)versions from the firmwareVersions
+        // parameter
+        firmwareVersionsToCheck.removeAll(latestfirmwareVersionsOfEachModuleTypeInHistory);
+
+        return firmwareVersionsToCheck;
+    }
+
+    // Helper class to keep track of InstallationDate and FirmwareVersion
+    private static class FirmwareVersionWithInstallationDate {
+        private final Date installationDate;
+        private final FirmwareVersion firmwareVersion;
+
+        public Date getInstallationDate() {
+            return this.installationDate;
+        }
+
+        public FirmwareVersion getFirmwareVersion() {
+            return this.firmwareVersion;
+        }
+
+        public FirmwareVersionWithInstallationDate(final Date installationDate, final FirmwareVersion firmwareVersion) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Move this constructor to comply with Java Code Conventions. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1213)